File: 高级数据结构与算法/C/第2章 分治法/2.3 快速幂/fast_exponentiation.c
#include <stdio.h>

/**
 * @brief  快速幂计算a^n
 */
int fastExp(int a, int n) {
    int result = 1;
    while(n) {
        if(n & 1) {
            result *= a;
        }
        a *= a;
        n >>= 1;
    }
    return result;
}

int main() {
    // 2^18 = 262144
    printf("%d\n", fastExp(2, 18));
    return 0;
}

File: 高级数据结构与算法/Java/第2章 分治法/2.3 快速幂/FastExponentiation.java
public class FastExponentiation {
    public static void main(String[] args) {
        // 2^18 = 262144
        System.out.println(fastExp(2, 18));
    }

    /**
     * 快速幂计算a^n
     */
    public static int fastExp(int a, int n) {
        int result = 1;
        while(n > 0) {
            if((n & 1) != 0) {
                result *= a;
            }
            a *= a;
            n >>= 1;
        }
        return result;
    }
}
